runtime/src/accounts_index.rs

```diff
@@ -1,38 +1,276 @@
+            .slot_list
+            .iter()
+            .enumerate()
+            .filter_map(|(i, (s, value))| {
+                if *s == slot {
+                    Some((i, (*s, value)))
+                } else {
+                    None
+                }
+            })
+            .collect();
+        assert!(same_slot_previous_updates.len() <= 1);
+
+        if let Some((list_index, (s, previous_update_value))) = same_slot_previous_updates.pop() {
+            reclaims.push((s, previous_update_value.clone()));
+            self.slot_list.remove(list_index);
+        } else {
+            // Only increment ref count if the account was not prevously updated in this slot
+            self.ref_count.fetch_add(1, Ordering::Relaxed);
+        }
+        self.slot_list.push((slot, account_info));
+    }
+}
+
 #[derive(Debug, Default)]
-pub struct AccountsIndex<T> {
-    pub account_maps: AccountMap<Pubkey, AccountMapEntry<T>>,
+pub struct RootsTracker {
+    roots: HashSet<Slot>,
+    uncleaned_roots: HashSet<Slot>,
+    previous_uncleaned_roots: HashSet<Slot>,
+}
+
+pub struct AccountsIndexIterator<'a, T> {
+    account_maps: &'a RwLock<AccountMap<Pubkey, AccountMapEntry<T>>>,
+    start_bound: Bound<Pubkey>,
+    end_bound: Bound<Pubkey>,
+}
+
+impl<'a, T> AccountsIndexIterator<'a, T> {
+    fn clone_bound(bound: Bound<&Pubkey>) -> Bound<Pubkey> {
+        match bound {
+            Unbounded => Unbounded,
+            Included(k) => Included(*k),
+            Excluded(k) => Excluded(*k),
+        }
+    }
 
-    pub roots: HashSet<Slot>,
-    pub uncleaned_roots: HashSet<Slot>,
-    pub previous_uncleaned_roots: HashSet<Slot>,
+    pub fn new<R>(
+        account_maps: &'a RwLock<AccountMap<Pubkey, AccountMapEntry<T>>>,
+        range: Option<R>,
+    ) -> Self
+    where
+        R: RangeBounds<Pubkey>,
+    {
+        Self {
+            start_bound: range
+                .as_ref()
+                .map(|r| Self::clone_bound(r.start_bound()))
+                .unwrap_or(Unbounded),
+            end_bound: range
+                .as_ref()
+                .map(|r| Self::clone_bound(r.end_bound()))
+                .unwrap_or(Unbounded),
+            account_maps,
+        }
+    }
 }
 
-impl<'a, T: 'a + Clone> AccountsIndex<T> {
-    fn do_scan_accounts<F, I>(&self, ancestors: &Ancestors, mut func: F, iter: I)
+impl<'a, T: 'static + Clone> Iterator for AccountsIndexIterator<'a, T> {
+    type Item = Vec<(Pubkey, AccountMapEntry<T>)>;
+    fn next(&mut self) -> Option<Self::Item> {
+        let chunk: Vec<(Pubkey, AccountMapEntry<T>)> = self
+            .account_maps
+            .read()
+            .unwrap()
+            .range((self.start_bound, self.end_bound))
+            .map(|(pubkey, account_map_entry)| {
+                (*pubkey, account_map_entry.clone())
+            })
+            .take(ITER_BATCH_SIZE)
+            .collect();
+
+        if chunk.is_empty() {
```

Comment:
nits: there is very and very small corner case of violating `iter` invariant: we could return `Some` even after we returned `None` here if we got newer Pubkey inserted, which would be compared after current `self.start_bound`. how about closing this hole now? Consider data parallelism from `rayon` and imagine how hard to spot this bug in the future from insane behavior. :)

Sadly, (start, end) = (unbound, unbound) or anything like that wouldn't work to mark the state. So, we must add another flag `finished` or like that and set it to `true` here.

Reply:
Fixed and added test!